File: world-edit/render_buffer.js
class RenderBuffer
{
    constructor()
    {
        this.vao;
        this.vbo;
        this.ebo;
        this.vertex_pos;
        this.index_pos;
        this.index_offset;
        this.vertices;
        this.indices;
    }
    static Init(gl, position, color, texture, vertex_limit, index_limit)
    {
        let buffer = new RenderBuffer();
        buffer.vertex_pos = 0;
        buffer.index_pos = 0;
        buffer.index_offset = 0;
        buffer.vertices = new Float32Array(vertex_limit * (position + color + texture));
        buffer.indices = new Uint16Array(index_limit);
        RenderSystem.MakeVao(gl, buffer, position, color, texture);
        return buffer;
    }
    static InitCopy(gl, source)
    {
        let buffer = new RenderBuffer(); 
        buffer.vertices = new Float32Array(source.vertex_pos);
        buffer.indices = new Uint16Array(source.index_pos);
        RenderBuffer.Copy(source, buffer);
        RenderSystem.MakeVao(gl, buffer, source.position, source.color, source.texture);
        RenderSystem.UpdateVao(gl, buffer);
        return buffer;
    }
    static Zero(buffer)
    {
        buffer.vertex_pos = 0;
        buffer.index_pos = 0;
        buffer.index_offset = 0;
    }
    static Copy(source, destination)
    {
        for (let i = 0; i < source.vertex_pos; i++)
        {
            destination.vertices[i] = source.vertices[i];
        }
        for (let i = 0; i < source.index_pos; i++)
        {
            destination.indices[i] = source.indices[i];
        }
        destination.index_offset = source.index_offset;
    }
}

File: source/graphics/renderbuffer.c
#include "renderbuffer.h"

renderbuffer *renderbuffer_init(int position, int color, int texture, int vertex_limit, int index_limit) {
    renderbuffer *r = safe_malloc(sizeof(renderbuffer));
    r->position = position;
    r->color = color;
    r->texture = texture;
    r->vertex_pos = 0;
    r->vertex_size = vertex_limit * (position + color + texture) * sizeof(GLfloat);
    r->index_pos = 0;
    r->index_size = index_limit * sizeof(GLuint);
    r->index_offset = 0;
    r->vertices = safe_malloc(r->vertex_size);
    r->indices = safe_malloc(r->index_size);
    return r;
}

void renderbuffer_zero(renderbuffer *self) {
    self->vertex_pos = 0;
    self->index_pos = 0;
    self->index_offset = 0;
}
